### PR TITLE
Refactor(plugins): Data validation tooling

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
## Change Summary

Refactor(plugins): Data validation tooling

Part of a multi-PR change.
This PR is just preparing tools to be used in later PRs.

## Proposed changes

- Retire `arista.avd.validation` plugin - validation moved to the action plugin `validate_and_template`
- Prepare new schematools to share common Ansible Action plugin code.
- Add support for ChainMap in type validation (accepted as dict)

### User Checklist

- N/A

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/plugins/plugin_utils/errors/errors.py [hunk 1]
```diff
@@ -50,3 +50,12 @@ def __init__(self, message: str = "Schema Error", error=None):
         else:
             self.message = message
         super().__init__(self.message)
+
+
+class AvdConversion(AristaAvdError):
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/errors/errors.py]: I would call it `AvdConversionWarning` for clarity / convention(?)

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        if validation_counter := self.handle_validation_exceptions(exceptions, self.validation_mode):
+            result_messages.append(f"{validation_counter} errors found during schema validation of input vars.")
+
+        return result_messages
+
+    def convert_and_validate_data(self, data):
+        """
+        Convert & Validate data according to the schema
+
+        Calls convertion and validation methods and gather resulting messages
```
Comment (JulioPDX) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]:
```suggestion
        Calls conversion and validation methods and gather resulting messages
```

Comment (JulioPDX) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]: Minor comment, besides that LGTM. Thank you for the great PR

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]: Thanks. Found the same in a few more places.

**Review by JulioPDX — APPROVED** [review]
LGTM

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            return []
+
+        result_messages = []
+        exceptions = self.avdschema.convert(data)
+        if conversion_counter := self.handle_validation_exceptions(exceptions, self.conversion_mode):
+            result_messages.append(f"{conversion_counter} data conversions done to conform to schema.")
+
+        return result_messages
+
+    def validate_data(self, data):
```
Comment (Shivani-chourasiya) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]:
```suggestion
    def validate_data(self, data) -> list:
```

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            return []
+
+        result_messages = []
+        exceptions = self.avdschema.validate(data)
+        if validation_counter := self.handle_validation_exceptions(exceptions, self.validation_mode):
+            result_messages.append(f"{validation_counter} errors found during schema validation of input vars.")
+
+        return result_messages
+
+    def convert_and_validate_data(self, data):
```
Comment (Shivani-chourasiya) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]:
```suggestion
    def convert_and_validate_data(self, data) -> dict:
```

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py (hunk 3 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]: Improved type hints in https://github.com/aristanetworks/ansible-avd/pull/2317/commits/72cc53e168325200fe4b0dafee3c66d21206608f

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py (hunk 4 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]: Improved type hints in https://github.com/aristanetworks/ansible-avd/pull/2317/commits/72cc53e168325200fe4b0dafee3c66d21206608f

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    def convert_and_validate_data(self, data: dict) -> dict:
+        """
+        Convert & Validate data according to the schema
+
+        Calls conversion and validation methods and gather resulting messages
+
+        Returns dict which can contain either or both of the following keys:
+        - failed: <bool>
+        - msg: <str with summarys messages from conversion and validation>
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]:
```suggestion
        - msg: <str> concatenation of the messages from conversion and validation
```

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+        The return value should be updated on Ansible Action "result"
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]: Maybe let's reword this

The return value should be included in the Ansible Action "result" (updated is very python dict)

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+            result_messages.extend(validation_messages)
+            if self.validation_mode == "error":
+                result["failed"] = True
+
+        if result_messages:
+            result["msg"] = " ".join(result_messages)
+
+        return result
+
+    def handle_validation_exceptions(self, exceptions: Generator, mode: str) -> int:
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py:126]:
```suggestion
    def handle_validation_exceptions(self, exceptions: Generator, mode: str) -> int:
        """
        Iterate through the Generator of exceptions.
        This method is actually where the content of the generator gets executed.
        
        It displays various messages depending on the `mode` parameter
        
        Returns:
        - counter: <int> the number of AristaAvdError in the exceptions Generator
        """
```

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py (hunk 5 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]: it is not all the messages, only the summary xx conversions was done or xx errors etc.

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py (hunk 6 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]: well it is specifically the "result" dicitonary which should be .update().

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py (hunk 7 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py:126]: Done in https://github.com/aristanetworks/ansible-avd/pull/2317/commits/dafa42ba59550da6873460b35d38219c1ed3ea07

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py (hunk 6 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]: Adjusted in https://github.com/aristanetworks/ansible-avd/pull/2317/commits/dafa42ba59550da6873460b35d38219c1ed3ea07

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py (hunk 5 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschematools.py]: Adjusted in https://github.com/aristanetworks/ansible-avd/pull/2317/commits/dafa42ba59550da6873460b35d38219c1ed3ea07

File: ansible_collections/arista/avd/plugins/plugin_utils/errors/errors.py (hunk 1 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/errors/errors.py]: Done in https://github.com/aristanetworks/ansible-avd/pull/2317/commits/dafa42ba59550da6873460b35d38219c1ed3ea07

**Review by gmuloc — APPROVED** [review]
LGTM

**Pull request merged**